compiler/typecheck/TcGenDeriv.hs

```diff
@@ -1955,7 +1955,7 @@ genAuxBindSpec dflags loc (DerivCon2Tag tycon)
     rdr_name = con2tag_RDR dflags tycon
 
     sig_ty = mkLHsSigWcType $ L loc $ XHsType $ NHsCoreTy $
-             mkSpecSigmaTy (tyConTyVars tycon) (tyConStupidTheta tycon) $
+             mkSpecSigmaTy (tyConTyVars tycon) [] $
```

Comment:
As part of the `deriving` mechanism, `genAuxBindSpec dflags loc (DerivCon2Tag tycon)` generates bindings of the form 

```haskell
$con2tag_<...> :: <ty> -> GHC.Prim.Int#
$con2tag_<...> x = GHC.Base.getTag x
```

as the name suggests, these convert a `con`structor into the corresponding `Int` `tag` (or rank). These bindings are used by the `derive`d instances for classes `Eq`, `Ord`, `Enum` and `Ix`* depending on some heuristics (e.g. it is only used by `Eq` if the type has more than 10 nullary constructors). For some reason, when the data type has a `DatatypeContext` (aka a "stupid theta"), it is used as the constraint of the corresponding `$con2tag_...` binding, even though it cannot be called directly by users and it has no impact on the type or behaviour of indirect calls. The corresponding logic in upstream GHC (now at `compiler/GHC/Tc/Deriv/Generate.hs`) has changed a bit, but notably the `$con2tag_...` bindings are no longer generated and `GHC.Prim.dataToTag#` is used instead. That is to say, I'm confident the change in this PR is safe.

This constraint was the cause of https://github.com/digital-asset/daml/issues/18504, since we add the `DamlEnum` constraint as the stupid theta for single-constructor enum declarations, but remove the `DamlEnum` class and instance in LF, so the LF version of `$con2tag_...` could not typecheck.

\* `Ix` doesn't exist in daml but the code is still present in this fork